File: dotnet/src/Ilmek.Core/Journal.cs
namespace Ilmek;

/// <summary>
/// A journal entry (MODEL.md §5).
///
/// An interrupt is a step whose value comes from a human (MODEL.md §6), so both
/// live in this one table: <c>Done</c> covers a completed step *and* an answered
/// interrupt; <c>Pending</c> is an interrupt still waiting on a person.
/// </summary>
public sealed record JournalEntry(bool Done, object? Value, object? Payload)
{
    public static JournalEntry AsDone(object? value) => new(true, value, null);
    public static JournalEntry AsPending(object? payload) => new(false, null, payload);
}

/// <summary>
/// The per-task record of completed steps and interrupt answers (MODEL.md §5).
///
/// This is the reason ilmek is not a LangGraph clone. A pure-replay engine
/// re-executes a node from the top on resume, so every side effect before the
/// pause happens twice. Here each effect is wrapped in a <c>ctx.StepAsync</c>
/// whose result is journaled — on replay the step returns the recorded value and
/// the function is never called again.
///
/// <para>A journal is scoped to a task and is dropped once that task's update is
/// reduced. It is replay memory, not history.</para>
/// </summary>
public sealed class Journal
{
    private readonly Dictionary<string, JournalEntry> _entries;
    private readonly List<string> _order;

    public Journal()
    {
        _entries = new Dictionary<string, JournalEntry>();
        _order = new List<string>();
    }

    private Journal(Dictionary<string, JournalEntry> entries, List<string> order)
    {
        _entries = entries;
        _order = order;
    }

    /// <summary>Rebuild from the persisted, order-significant form.</summary>
    public static Journal Load(IEnumerable<KeyValuePair<string, JournalEntry>> dump)
    {
        var journal = new Journal();
        foreach (var (key, entry) in dump) journal.Put(key, entry);
        return journal;
    }

    /// <summary>The persisted form: entries in journaled order.</summary>
    public IReadOnlyList<KeyValuePair<string, JournalEntry>> Dump() =>
        _order.Select(k => new KeyValuePair<string, JournalEntry>(k, _entries[k])).ToList();

    public IReadOnlyList<string> Keys => _order;

    public JournalEntry? Fetch(string key) => _entries.TryGetValue(key, out var e) ? e : null;

    public void Put(string key, JournalEntry entry)
    {
        if (!_entries.ContainsKey(key)) _order.Add(key);
        _entries[key] = entry;
    }

    public void PutDone(string key, object? value) => Put(key, JournalEntry.AsDone(value));

    public void PutPending(string key, object? payload) => Put(key, JournalEntry.AsPending(payload));

    /// <summary>Resolve a pending interrupt with the human's answer.</summary>
    public (bool Ok, string Reason) Answer(string key, object? value)
    {
        if (!_entries.TryGetValue(key, out var entry)) return (false, "unknown_key");
        if (entry.Done) return (false, "already_answered");
        Put(key, JournalEntry.AsDone(value));
        return (true, "");
    }

    /// <summary>Every interrupt still waiting on a human, in journaled order.</summary>
    public IReadOnlyList<(string Key, object? Payload)> Pending() =>
        _order.Where(k => !_entries[k].Done).Select(k => (k, _entries[k].Payload)).ToList();

    public Journal Clone() => new(new Dictionary<string, JournalEntry>(_entries), new List<string>(_order));
}

/// <summary>
/// Per-task bookkeeping around a journal: occurrence counters for key suffixing
/// and the trace strict mode checks (MODEL.md §5.4, §5.5).
///
/// A fresh instance per attempt resets the counters and trace while the
/// underlying <see cref="Journal"/> persists — which is what makes a retry
/// replay completed steps instead of re-running them (MODEL.md §16).
/// </summary>
public sealed class TaskJournal
{
    public Journal Journal { get; }
    private readonly Dictionary<string, int> _counters = new();
    private readonly List<string> _observed = new();

    public TaskJournal(Journal journal) => Journal = journal;

    /// <summary>
    /// Turn a caller-supplied base key into the full journal key by appending
    /// this pass's occurrence count: <c>"charge"</c> → <c>charge#0</c>,
    /// <c>charge#1</c>, …
    ///
    /// <para>Suffixing is uniform so the mapping depends only on how many times
    /// *this* base has been requested in *this* pass. A node sees identical state
    /// on every pass, so the sequence is stable.</para>
    /// </summary>
    public string ResolveKey(string baseKey)
    {
        var n = _counters.TryGetValue(baseKey, out var count) ? count : 0;
        _counters[baseKey] = n + 1;
        var full = $"{baseKey}#{n}";
        _observed.Add(full);
        return full;
    }

    /// <summary>
    /// Strict-mode check (MODEL.md §5.5): every key the journal holds must have
    /// been requested again on this pass. A key that vanishes means the node took
    /// a different path than before — a side effect escaped a step.
    /// </summary>
    public void CheckDeterminism(string node)
    {
        var seen = new HashSet<string>(_observed);
        var missing = Journal.Keys.Where(k => !seen.Contains(k)).ToList();
        if (missing.Count == 0) return;

        throw new NondeterminismException(
            $"node \"{node}\" replayed without requesting journaled step(s) [{string.Join(", ", missing)}].\n\n" +
            "The node body is not deterministic modulo steps: given the same state and journal it took a " +
            "different path. Wrap every side effect and every nondeterministic read (clock, RNG, guid, " +
            "network) in ctx.StepAsync — see MODEL.md §5.2.");
    }
}

File: ts/packages/core/src/journal.ts
import { NondeterminismError } from "./errors.ts";

/**
 * A journal entry (MODEL.md §5).
 *
 * An interrupt is a step whose value comes from a human (MODEL.md §6), so both
 * live in this one table: `done` covers a completed step *and* an answered
 * interrupt; `pending` is an interrupt still waiting on a person.
 */
export type JournalEntry =
    | { readonly status: "done"; readonly value: unknown }
    | { readonly status: "pending"; readonly payload: unknown };

export type JournalDump = ReadonlyArray<readonly [string, JournalEntry]>;

export interface PendingEntry {
    readonly key: string;
    readonly payload: unknown;
}

/**
 * The per-task record of completed steps and interrupt answers (MODEL.md §5).
 *
 * This is the reason ilmek is not a LangGraph clone. A pure-replay engine
 * re-executes a node from the top on resume, so every side effect before the
 * pause happens twice. Here each effect is wrapped in a `ctx.step()` whose
 * result is journaled — on replay the step returns the recorded value and the
 * function is never called again.
 *
 * A journal is scoped to a task (`{thread, checkpoint, node}`) and is dropped
 * once that task's update is reduced. It is replay memory, not history.
 */
export class Journal {
    // A Map preserves insertion order, so journaled order comes for free.
    private readonly entries: Map<string, JournalEntry>;

    constructor(entries: Map<string, JournalEntry> = new Map()) {
        this.entries = entries;
    }

    static load(dump: JournalDump): Journal {
        return new Journal(new Map(dump.map(([k, e]) => [k, e])));
    }

    dump(): JournalDump {
        return [...this.entries.entries()];
    }

    keys(): string[] {
        return [...this.entries.keys()];
    }

    fetch(key: string): JournalEntry | undefined {
        return this.entries.get(key);
    }

    putDone(key: string, value: unknown): void {
        this.entries.set(key, { status: "done", value });
    }

    putPending(key: string, payload: unknown): void {
        this.entries.set(key, { status: "pending", payload });
    }

    /** Resolve a pending interrupt with the human's answer. */
    answer(key: string, value: unknown): { ok: true } | { ok: false; reason: string } {
        const entry = this.entries.get(key);
        if (!entry) return { ok: false, reason: "unknown_key" };
        if (entry.status === "done") return { ok: false, reason: "already_answered" };
        this.entries.set(key, { status: "done", value });
        return { ok: true };
    }

    /** Every interrupt still waiting on a human, in journaled order. */
    pending(): PendingEntry[] {
        const out: PendingEntry[] = [];
        for (const [key, entry] of this.entries) {
            if (entry.status === "pending") out.push({ key, payload: entry.payload });
        }
        return out;
    }
}

/**
 * Per-task bookkeeping around a journal: occurrence counters for key
 * suffixing and the trace strict mode checks (MODEL.md §5.4, §5.5).
 *
 * In Elixir this state lives in the task process's dictionary; in TypeScript a
 * task is just an async call tree, so it lives on this object — which the
 * task's `ctx` closes over. Same lifetime, no ambient storage required.
 */
export class TaskJournal {
    readonly journal: Journal;
    private readonly counters = new Map<string, number>();
    private readonly observed: string[] = [];

    constructor(journal: Journal) {
        this.journal = journal;
    }

    /**
     * Turn a caller-supplied base key into the full journal key by appending
     * this pass's occurrence count for that base: `"charge"` → `"charge#0"`,
     * `"charge#1"`, …
     *
     * Suffixing is uniform (even for keys used once) so the mapping depends only
     * on how many times *this* base has been requested in *this* pass. Since a
     * node sees identical state on every pass the sequence is stable, and a node
     * that branches around a step only shifts the ordinals of that same base key.
     */
    resolveKey(base: string): string {
        const n = this.counters.get(base) ?? 0;
        this.counters.set(base, n + 1);
        const full = `${base}#${n}`;
        this.observed.push(full);
        return full;
    }

    /**
     * Strict-mode check (MODEL.md §5.5): every key the journal holds must have
     * been requested again on this pass. A key that vanishes means the node took
     * a different path than it did before — a side effect escaped a step.
     */
    checkDeterminism(node: string): void {
        const seen = new Set(this.observed);
        const missing = this.journal.keys().filter((k) => !seen.has(k));
        if (missing.length === 0) return;

        throw new NondeterminismError(
            `node ${JSON.stringify(node)} replayed without requesting journaled step(s) ` +
                `${JSON.stringify(missing)}.\n\n` +
                `The node body is not deterministic modulo steps: given the same state and ` +
                `journal it took a different path. Wrap every side effect and every ` +
                `nondeterministic read (clock, RNG, uuid, network) in ctx.step() — see ` +
                `MODEL.md §5.2.`,
        );
    }
}
